File: leafdb.py
import requests
from typing import Any, Callable
from fastapi import HTTPException
from functools import wraps
from json import dumps, loads


class Leafdb:

    def __init__(self):
        self.__url = "http://localhost:8000"

    def set_port(self, port: int) -> Any:
        """Sets the port to the URL of the database"""
        if not isinstance(port, int):
            raise ValueError(f"Invalid value: {port}")
        self.__url = f"http://localhost:{port}"

    def view(self, key: Any = None) -> dict | None:
        """Views database items.
        key -- If given, it's value from the database is returned, otherwise returns all the database
        """
        if key is None:
            res = requests.get(self.__url)
            return res.json()

        res = requests.get(f"{self.__url}/{key}")

        return res.json()

    def insert(self, data: dict) -> None:
        """Inserts <data> dictionary object to the database"""
        if not isinstance(data, dict):
            raise TypeError(f"invalid data structure: \n {data}")
        requests.post(self.__url, json=data)

    def delete(self, *items: list[Any]) -> None:
        """Deletes <items> from the database"""
        for item in items:
            requests.delete(f"{self.__url}/{item}")

    def edit(self, key: Any, value: Any) -> None:
        """Changes `key`'s value to `value` in the database"""
        requests.put(
            f"{self.__url}/{key}",
            json={"value": value}
        )

    def get_url(self) -> str:
        return self.__url


__all__ = ["Leafdb"]

File: leafdb.js
class Leafdb {
  #url; // (encapsulating url)

  constructor() {
    this.#url = `http://localhost:8000`;
  }
  /* Sets the port used in the uvicorn server */
  setPort(port) {
    if (!(typeof port === "number")) {
      throw new SyntaxError(`Invalid port: ${port}`);
    }
    this.#url = `http://localhost:${port}`;
  }
  /**
   * Returns an object with the database data
   * @param {number} data optional argument, if specified, returns that specific item, otherwise returns all the database
   */
  async view(data = null) {
    if (data) {
      const rawContent = await fetch(`${this.#url}/${data}`);
      if (!rawContent.ok) {
        throw new Error(`${rawContent.status} ${rawContent.statusText}`);
      }
      const content = await rawContent.json();
      return content;
    }

    const rawContent = await fetch(this.#url);
    if (!rawContent.ok) {
      throw new Error(`${rawContent.status} ${rawContent.statusText}`);
    }
    const content = await rawContent.json();
    return content;
  }
  /* Deletes `item` from the database */
  async del(item) {

    const data = await new Leafdb().view();
    if (!Object.keys(data).includes(item)) {
      console.warn(`${item} isn't found in database`);
      return;
    }

    const res = await fetch(`${this.#url}/${item}`, {
      method: "DELETE",
    });

    if (!res.ok) {
      throw new Error(`${res.status} ${res.statusText}`);
    }
  }
  /* Inserts `data` to the database */
  async insert(data) {
    if (!(data instanceof Object) || data === null || Array.isArray(data)) {
      throw new TypeError(`Invalid structure: ${data}`);
    }

    const res = await fetch(this.#url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    });

    if (!res.ok) {
      throw new Error(`${res.status} ${res.statusText}`);
    }
  }
  /* Changes `key`'s value to `value` in the database */
  async edit(key, value) {
    if (key === null || Array.isArray(key)) {
      throw new TypeError(`Invalid structure: ${key}`);
    }

    const res = await fetch(`${this.#url}/${key}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ value }),
    });

    if (!res.ok) {
      throw new Error(`${res.status} ${res.statusText}`);
    }
  }

  /* Provides read-only access to the URL */
  getUrl() {
    return this.#url;
  }
}
export default Leafdb;
